federatedscope/core/trainers/context.py

```diff
@@ -159,14 +182,21 @@ def pre_calculate_batch_epoch_num(self, local_update_steps):
         return num_train_batch, num_train_batch_last_epoch, num_train_epoch, num_total_train_batch
 
     def append_mode(self, mode):
-        self.mode.append(mode)
-        self.cur_mode = self.mode[-1]
+        if mode in self.mode_stack:
```

Comment:
What is the scope of valid modes? According to this code snippet, there are n! modes (suppose there are n choices to be appended)

Reply:
Generally, we choose mode from "train", "test" and "val", which corresponds to the different running processes in FL/ML. 
Actually, only the final mode in `self.mode` represents the current mode, and we track the previous ones to support  nested running routines (e.g. running evaluation within the training process) and recover the status of the learning model (`model.train/eval()`) after jumping out of the nested routine.